### PR TITLE
Fixes race condition in MatrixWorkspace::rebuildDetectorIDGroupings

### PR DESCRIPTION
**Pull request by martyngigg**
Description of work.

See commit message.

**To test:**

* Code review and ensure the tests pass.

Refs #22063 

**Release Notes** 
*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/MatrixWorkspace.cpp [hunk 1]
```diff
@@ -2014,22 +2014,29 @@ void MatrixWorkspace::rebuildDetectorIDGroupings() {
...
+        errorCode = 2;
       } else {
         detIDs.insert(allDetIDs[detIndex]);
       }
     }
     spec.setDetectorIDs(std::move(detIDs));
   }
-  if (parallelException)
-    throw std::invalid_argument(error);
+  if (parallelException) {
```
Comment (SimonHeybrock) [inline Framework/API/src/MatrixWorkspace.cpp]: Can we remove `parallelException`? Check for non-zero `errorCode` enough?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Code looks good, but according to one of the related tests a wrong exception is thrown now. I had a look an cannot see why right now.

See comment below for possible cleanup.

File: Framework/API/src/MatrixWorkspace.cpp (hunk 1 above)
Comment (martyngigg) [reply Framework/API/src/MatrixWorkspace.cpp]: Good point - that should work.

**Comment by SimonHeybrock** [comment]
@quantumsteve If I am not mistaken, the parallel macros work only within algorithms and that is the reason for the hand-written OMP loop here.

**Review by SimonHeybrock — APPROVED** [review]

**Comment by martyngigg** [comment]
The usual OMP macros expect 2 private member variables of Algorithm to be accessible - I'm not sure why the first macro can't just define them locally though. The main difference here is that an exception isn't being caught and rethrown we're just setting a flag to say what should be thrown later. The macros have a `try...catch` block to catch anything so I don't think we'd gain much in those cases.

**Comment by martyngigg** [comment]
@quantumsteve I've made a small update to use an enum class rather than an `int` as you suggested to given some meaning to the codes.

**Pull request merged**